File: taurus-protect-sdk-python/taurus_protect/crypto/keys.py
"""Key handling utilities for Taurus-PROTECT SDK."""

from __future__ import annotations

from typing import List, Optional

from cryptography.exceptions import UnsupportedAlgorithm
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.primitives.asymmetric.ec import (
    EllipticCurvePrivateKey,
    EllipticCurvePublicKey,
    SECP256R1,
)


def decode_public_key_pem(pem_data: str) -> EllipticCurvePublicKey:
    """
    Decode a PEM-encoded ECDSA public key.

    Only P-256 (secp256r1) curve is supported for security reasons.

    Args:
        pem_data: PEM-encoded public key string.

    Returns:
        ECDSA public key object (P-256 curve).

    Raises:
        ValueError: If the key cannot be decoded, is not an EC key,
                   or uses an unsupported curve.

    Example:
        >>> key = decode_public_key_pem('''-----BEGIN PUBLIC KEY-----
        ... MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE...
        ... -----END PUBLIC KEY-----''')
    """
    try:
        key = serialization.load_pem_public_key(pem_data.encode("utf-8"))
        if not isinstance(key, EllipticCurvePublicKey):
            raise ValueError("Key is not an elliptic curve public key")
        # Validate curve is P-256 (secp256r1) - reject weaker curves
        if not isinstance(key.curve, SECP256R1):
            raise ValueError(
                f"Only P-256 (secp256r1) curve is supported, got {key.curve.name}"
            )
        return key
    except ValueError as e:
        # Wrap cryptography library errors to provide consistent messaging
        raise ValueError(f"Failed to decode public key: {e}") from e
    except (TypeError, UnsupportedAlgorithm) as e:
        raise ValueError(f"Failed to decode public key: {e}") from e


def decode_public_keys_pem(pem_keys: List[str]) -> List[EllipticCurvePublicKey]:
    """
    Decode multiple PEM-encoded public keys.

    Args:
        pem_keys: List of PEM-encoded public key strings.

    Returns:
        List of ECDSA public key objects.
    """
    return [decode_public_key_pem(pem) for pem in pem_keys]


def decode_private_key_pem(
    pem_data: str,
    password: Optional[bytes] = None,
) -> EllipticCurvePrivateKey:
    """
    Decode a PEM-encoded ECDSA private key.

    Supports both EC PRIVATE KEY and PKCS#8 formats.
    Only P-256 (secp256r1) curve is supported for security reasons.

    Args:
        pem_data: PEM-encoded private key string.
        password: Optional password for encrypted keys.

    Returns:
        ECDSA private key object (P-256 curve).

    Raises:
        ValueError: If the key cannot be decoded, is not an EC key,
                   or uses an unsupported curve.

    Example:
        >>> key = decode_private_key_pem('''-----BEGIN EC PRIVATE KEY-----
        ... MHQCAQEEIDfN...
        ... -----END EC PRIVATE KEY-----''')
    """
    try:
        key = serialization.load_pem_private_key(
            pem_data.encode("utf-8"),
            password=password,
        )
        if not isinstance(key, EllipticCurvePrivateKey):
            raise ValueError("Key is not an elliptic curve private key")
        # Validate curve is P-256 (secp256r1) - reject weaker curves
        if not isinstance(key.curve, SECP256R1):
            raise ValueError(
                f"Only P-256 (secp256r1) curve is supported, got {key.curve.name}"
            )
        return key
    except ValueError as e:
        # Wrap cryptography library errors to provide consistent messaging
        raise ValueError(f"Failed to decode private key: {e}") from e
    except (TypeError, UnsupportedAlgorithm) as e:
        raise ValueError(f"Failed to decode private key: {e}") from e


def encode_public_key_pem(key: EllipticCurvePublicKey) -> str:
    """
    Encode an ECDSA public key to PEM format.

    Args:
        key: ECDSA public key object.

    Returns:
        PEM-encoded public key string.
    """
    pem_bytes = key.public_bytes(
        encoding=serialization.Encoding.PEM,
        format=serialization.PublicFormat.SubjectPublicKeyInfo,
    )
    return pem_bytes.decode("utf-8")

File: taurus-protect-sdk-typescript/src/crypto/keys.ts
/**
 * Key handling utilities for Taurus-PROTECT SDK.
 *
 * Provides PEM key parsing for ECDSA P-256 keys.
 */

import * as crypto from "crypto";

/**
 * Decode a PEM-encoded ECDSA private key.
 *
 * Supports both EC PRIVATE KEY and PKCS#8 formats.
 *
 * @param pem - PEM-encoded private key string
 * @returns Node.js KeyObject for the private key
 * @throws Error if the key cannot be decoded or is not an EC key
 *
 * @example
 * ```typescript
 * const privateKey = decodePrivateKeyPem(`-----BEGIN EC PRIVATE KEY-----
 * MHQCAQEEIDfN...
 * -----END EC PRIVATE KEY-----`);
 * ```
 */
export function decodePrivateKeyPem(pem: string): crypto.KeyObject {
  try {
    const key = crypto.createPrivateKey({
      key: pem,
      format: "pem",
    });

    // Verify it's an EC key
    const keyDetails = key.asymmetricKeyDetails;
    if (!keyDetails || keyDetails.namedCurve !== "prime256v1") {
      // Also accept P-256 alias
      if (!keyDetails || keyDetails.namedCurve !== "P-256") {
        throw new Error(
          `Expected P-256/prime256v1 EC key, got: ${keyDetails?.namedCurve ?? "unknown"}`
        );
      }
    }

    return key;
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Failed to decode private key: ${error.message}`);
    }
    throw new Error("Failed to decode private key: unknown error");
  }
}

/**
 * Decode a PEM-encoded ECDSA public key.
 *
 * @param pem - PEM-encoded public key string
 * @returns Node.js KeyObject for the public key
 * @throws Error if the key cannot be decoded or is not an EC key
 *
 * @example
 * ```typescript
 * const publicKey = decodePublicKeyPem(`-----BEGIN PUBLIC KEY-----
 * MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAE...
 * -----END PUBLIC KEY-----`);
 * ```
 */
export function decodePublicKeyPem(pem: string): crypto.KeyObject {
  try {
    const key = crypto.createPublicKey({
      key: pem,
      format: "pem",
    });

    // Verify it's an EC key
    const keyDetails = key.asymmetricKeyDetails;
    if (!keyDetails || keyDetails.namedCurve !== "prime256v1") {
      // Also accept P-256 alias
      if (!keyDetails || keyDetails.namedCurve !== "P-256") {
        throw new Error(
          `Expected P-256/prime256v1 EC key, got: ${keyDetails?.namedCurve ?? "unknown"}`
        );
      }
    }

    return key;
  } catch (error) {
    if (error instanceof Error) {
      throw new Error(`Failed to decode public key: ${error.message}`);
    }
    throw new Error("Failed to decode public key: unknown error");
  }
}

/**
 * Decode multiple PEM-encoded public keys.
 *
 * @param pems - Array of PEM-encoded public key strings
 * @returns Array of Node.js KeyObject for the public keys
 * @throws Error if any key cannot be decoded
 *
 * @example
 * ```typescript
 * const publicKeys = decodePublicKeysPem([pem1, pem2, pem3]);
 * ```
 */
export function decodePublicKeysPem(pems: string[]): crypto.KeyObject[] {
  return pems.map((pem) => decodePublicKeyPem(pem));
}

/**
 * Encode a public key to PEM format.
 *
 * @param key - Node.js KeyObject for the public key
 * @returns PEM-encoded public key string
 */
export function encodePublicKeyPem(key: crypto.KeyObject): string {
  return key.export({ type: "spki", format: "pem" }) as string;
}

/**
 * Extract the public key from a private key.
 *
 * @param privateKey - Node.js KeyObject for the private key
 * @returns Node.js KeyObject for the corresponding public key
 */
export function getPublicKeyFromPrivate(
  privateKey: crypto.KeyObject
): crypto.KeyObject {
  return crypto.createPublicKey(privateKey);
}
